### PR TITLE
Update access_token procedure

### PR DESCRIPTION
**Pull request by eggegg31415**
* Add access token request scope and NF service consumre authentication verification
* Change NRF access token signing method from constant secret to RSA asynchronous encryption

### CODE REVIEW

**Comment by free5gc-org** [comment]
@eggegg31415 

Please rebase this PR with the latest commit.
Some packages are already renamed, so please help fix the conflicts with new packages and APIs.
Thanks.

File: factory/config.go [hunk 1]
```diff
@@ -35,11 +35,12 @@ type Info struct {
...
-	MongoDBName     string        `yaml:"MongoDBName"`
-	MongoDBUrl      string        `yaml:"MongoDBUrl"`
-	DefaultPlmnId   models.PlmnId `yaml:"DefaultPlmnId"`
-	ServiceNameList []string      `yaml:"serviceNameList,omitempty"`
+	Sbi             *Sbi               `yaml:"sbi,omitempty"`
+	MongoDBName     string             `yaml:"MongoDBName"`
+	MongoDBUrl      string             `yaml:"MongoDBUrl"`
+	DefaultPlmnId   models.PlmnId      `yaml:"DefaultPlmnId"`
+	ServiceNameList []string           `yaml:"serviceNameList,omitempty"`
+	ServiceList     []models.NfService `yaml:"serviceList,omitempty"`
```
Comment (free5gc-org) [inline factory/config.go:43]: Why to add ServiceList? You didn't use the new config.

File: factory/config.go [hunk 2]
```diff
@@ -48,6 +49,7 @@ type Sbi struct {
 	// IPv6Addr  string `yaml:"ipv6Addr,omitempty"`
 	BindingIPv4 string `yaml:"bindingIPv4,omitempty"` // IP used to run the server in the node.
 	Port        int    `yaml:"port,omitempty"`
+	OAuth       bool   `yaml:"oauth,omitempty"`
```
Comment (free5gc-org) [inline factory/config.go:52]: Why to add OAuth? You didn't use the new config.

File: producer/access_token.go [hunk 3]
```diff
@@ -53,13 +65,32 @@ func AccessTokenProcedure(request models.AccessTokenReq) (response *models.Acces
 	}
 	accessTokenClaims.IssuedAt = int64(now)
 
-	mySigningKey := []byte("NRF") // AllYourBase
-	token := jwt.NewWithClaims(jwt.SigningMethodHS256, accessTokenClaims)
-	accessToken, err := token.SignedString(mySigningKey)
+	// Use RSA as a signing method
+	privKeyPath := "../support/TLS/nrf.key"
```
Comment (free5gc-org) [inline producer/access_token.go:69]: Use `factory.NrfConfig.TLSKeyPath()` to get key path.

File: internal/sbi/producer/access_token.go [hunk 4]
```diff
@@ -74,3 +106,160 @@ func AccessTokenProcedure(request models.AccessTokenReq) (response *models.Acces
...
+	}
+
+	if strings.ToUpper(string(nfProfile.NfType)) != reqNfType {
+		errResponse = &models.AccessTokenErr{
+			Error: "invalid_client",
+		}
+		return errResponse
+	}
+	// Check with certificate
+	rootPEM, err := ioutil.ReadFile("../support/TLS/root.crt")
```
Comment (free5gc-org) [inline internal/sbi/producer/access_token.go:159]: move this file to `config/TLS/root.crt`

File: pkg/factory/config.go [hunk 5]
```diff
@@ -133,8 +135,10 @@ func (s *Sbi) validate() (bool, error) {
 }
 
 type Tls struct {
-	Pem string `yaml:"pem,omitempty" valid:"type(string),minstringlength(1),required"`
-	Key string `yaml:"key,omitempty" valid:"type(string),minstringlength(1),required"`
+	Pem     string `yaml:"pem,omitempty" valid:"type(string),minstringlength(1),required"`
+	Key     string `yaml:"key,omitempty" valid:"type(string),minstringlength(1),required"`
+	BasePem string `yaml:"basepem,omitempty" valid:"type(string),minstringlength(1),required"`
```
Comment (free5gc-org) [inline pkg/factory/config.go:140]: Replace `Base` with `Root`

File: pkg/factory/config.go [hunk 6]
```diff
@@ -225,3 +229,17 @@ func (c *Config) TLSKeyPath() string {
 	}
 	return NrfDefaultTLSKeyPath
 }
+
+func (c *Config) TLSPemBasePath() string {
```
Comment (free5gc-org) [inline pkg/factory/config.go:233]: Rename to `RootCertPemPath()`

File: pkg/factory/config.go [hunk 7]
```diff
@@ -225,3 +229,17 @@ func (c *Config) TLSKeyPath() string {
...
+	if c.Configuration.Sbi.Tls != nil {
+		return c.Configuration.Sbi.Tls.BasePem
+	}
+	return NrfDefaultTLSPemBasePath
+}
+
+func (c *Config) TLSKeyBasePath() string {
```
Comment (free5gc-org) [inline pkg/factory/config.go:240]: Rename to `RootCertKeyPath()`

File: pkg/factory/config.go [hunk 8]
```diff
@@ -114,15 +116,21 @@ type Sbi struct {
...
 		return str == "https" || str == "http"
 	})
 
-	if tls := s.Tls; tls != nil {
+	if tls := s.TlsCert; tls != nil {
+		if result, err := tls.validate(); err != nil {
+			return result, err
+		}
+	}
+	if tls := s.RootCert; tls != nil {
```
Comment (free5gc-org) [inline pkg/factory/config.go:133]: `if rc := s.RootCert; rc != nil {`

File: pkg/factory/config.go [hunk 9]
```diff
@@ -18,14 +18,16 @@ import (
...
 const (
-	NrfDefaultTLSPemPath     = "./config/TLS/nrf.pem"
-	NrfDefaultTLSKeyPath     = "./config/TLS/nrf.key"
-	NrfExpectedConfigVersion = "1.0.1"
-	NrfSbiDefaultIPv4        = "127.0.0.10"
-	NrfSbiDefaultPort        = 8000
-	NrfSbiDefaultScheme      = "https"
-	NrfNfmResUriPrefix       = "/nnrf-nfm/v1"
-	NrfDiscResUriPrefix      = "/nnrf-disc/v1"
+	NrfDefaultTLSPemPath      = "./config/TLS/nrf.pem"
```
Comment (free5gc-org) [inline pkg/factory/config.go:21]:
```
	NrfDefaultTLSPemPath      = "./config/cert/nrf.pem"
	NrfDefaultTLSKeyPath      = "./config/cert/nrf.key"
	NrfDefaultRootCertPemPath = "./config/cert/root.pem"
	NrfDefaultRootCertKeyPath = "./config/cert/root.key"
```

File: go.mod [hunk 10]
```diff
@@ -5,6 +5,8 @@ go 1.14
 require (
 	github.com/antonfisher/nested-logrus-formatter v1.3.1
 	github.com/asaskevich/govalidator v0.0.0-20210307081110-f21760c49a8d
+	github.com/free5gc/http_wrapper v1.0.0 // indirect
```
Comment (free5gc-org) [inline go.mod:8]: remove the 2 pkg:
```
	github.com/free5gc/http_wrapper v1.0.0 // indirect
	github.com/free5gc/logger_util v1.0.0
```
and then execute:
`go mod tidy`

File: internal/sbi/accesstoken/api_access_token_request.go [hunk 11]
```diff
@@ -25,9 +27,47 @@ import (
...
+		var vt reflect.Type
+		for i := 0; i < rt.NumField(); i++ {
+			if tag := rt.Field(i).Tag.Get("yaml"); tag == key {
+				name = rt.Field(i).Name
+				vt = rt.Field(i).Type
+				break
+			}
+		}
+		if vt.Name() == "string" || vt.Name() == "NfType" {
+			reflect.ValueOf(&accessTokenReq).Elem().FieldByName(name).SetString(value[0])
```
Comment (free5gc-org) [inline internal/sbi/accesstoken/api_access_token_request.go:50]: remove the empty line

File: internal/sbi/producer/access_token.go [hunk 12]
```diff
@@ -74,3 +106,172 @@ func AccessTokenProcedure(request models.AccessTokenReq) (response *models.Acces
...
+	rootCertPemPath := factory.NrfConfig.RootCertPemPath()
+	rootPEM, err := ioutil.ReadFile(rootCertPemPath)
+	if err != nil {
+		logger.AccessTokenLog.Errorln("Certificate verify error: " + err.Error())
+		errResponse = &models.AccessTokenErr{
+			Error: "invalid_client",
+		}
+		return errResponse
+	}
+	nfPemPath := rootCertPemPath[0:strings.Index(rootCertPemPath, "root.pem")]
```
Comment (free5gc-org) [inline internal/sbi/producer/access_token.go:169]:
```
nfPemDir, _ := filepath.Split(c.NrfCertPemPath())
certPEM, err := ioutil.ReadFile(nfPemDir + strings.ToLower(reqNfType) + ".pem")
```

File: internal/sbi/producer/access_token.go [hunk 13]
```diff
@@ -53,9 +66,28 @@ func AccessTokenProcedure(request models.AccessTokenReq) (response *models.Acces
 	}
 	accessTokenClaims.IssuedAt = int64(now)
 
-	mySigningKey := []byte("NRF") // AllYourBase
-	token := jwt.NewWithClaims(jwt.SigningMethodHS256, accessTokenClaims)
-	accessToken, err := token.SignedString(mySigningKey)
+	// Use RSA as a signing method
+	privKeyPath := factory.NrfConfig.TLSKeyPath()
```
Comment (free5gc-org) [inline internal/sbi/producer/access_token.go:70]: signBytes, err := ioutil.ReadFile(c.NrfCertKeyPath())

**Review by free5gc-org — APPROVED** [review]

**Comment by swthorn** [comment]
Hi,

Can you please describe how the TLS certificates are generated? I wrote a small bash script to do it:
```# !/bin/bash
NFS=("nrf" "amf" "ausf" "smf" "n3iwf" "pcf" "udm" "udr" "nssf")

openssl genrsa -out root.key 2048
openssl req -x509 -new -nodes -key root.key -sha256 -days 1825 -out root.pem

for i in "${NFS[@]}"
do
   openssl genrsa -out $i.key 2048
   openssl req -new -key $i.key -subj "/C=US/ST=NC/O=free5gc Inc./CN=$i" -addext "subjectAltName = DNS:$i,URI:$i.free5gc.org:NF_INSTANCE_ID_HERE1:NF_INSTANCE_ID_HERE2" -out $i.pem
   openssl x509 -req -extfile <(printf "subjectAltName=DNS:$i,URI:$i.free5gc.org:NF_INSTANCE_ID_HERE1:NF_INSTANCE_ID_HERE2") -in $i.pem -CA root.pem -CAkey root.key -CAcreateserial -out $i.pem -days 500 -sha256 
done
```

However, lines
https://github.com/free5gc/nrf/blob/518919ac94aaf59e3e5cbd70531cabd74919235f/internal/sbi/producer/access_token.go#L198-L206

specify that the NF Instance ID needs to be in the TLS certificate. As far as I can tell, in all NFs (amf, ausf, udr, ...) the nf instance ID is determined at runtime. How can this NF Instance ID be embedded in the TLS certificate without changing the certificate for every NF launch?

This is the error I am seeing:

> nrf              | 2022-09-01T17:02:33Z [ERRO][NRF][Token] Certificate verify error: NF Instance Id mismatch (Expected ID: f287408f-351d-4e31-9be1-6dfa73a7330f Received ID: NF_INSTANCE_ID_HERE2) (/go/src/free5gc/NFs/nrf/internal/sbi/producer/access_token.go:201 github.com/free5gc/nrf/internal/sbi/producer.AccessTokenScopeCheck)

**Comment by tim-ywliu** [comment]
@eggegg31415 could you help reply @swthorn 's question.
Thanks.

**Pull request merged**